### PR TITLE
Weak var runtime mirror 27348445

### PR DESCRIPTION
**Pull request by bitjammer**
94a9c51 made some changes to loading
weak references by adding some information in the lower bits with
respect to locking. These bits need to be masked out when performing a
load, such as when we want to get the metadata pointer for a class
instance. This normally works fine when going through the normal weak
loading functions in the runtime.

When the runtime function swift_ClassMirror_subscript gets the offset of
one of its stored properties, it immediately packages it into the the
ad-hoc existential container, known as just `Mirror` in the runtime.
However, the weak reference isn't aligned! It has bit 1 set.  We weren't
loading the weak reference here as we would during normal SILGen, such
as with a weak_load instruction. Simulate that here and make the
reference strong before putting it into the Mirror container, which also
clears those lower bits.

rdar://problem/27475034

There are still a couple of other cases to handle, namely the
unowned(safe) and unowned(unsafe) reference kinds. There may be other
places where an unaligned pointer is problematic in the runtime, which
we should audit for correctness.

rdar://problem/27809991

### CODE REVIEW

**Comment by bitjammer** [comment]
@swift-ci Please test.

**Comment by bitjammer** [comment]
@jckarter or @rjmccall can you review this change for Swift 3? I only fixed the weak case for now.

**Comment by jckarter** [comment]
Looks reasonable to me.

**Comment by bitjammer** [comment]
@swift-ci Please test.

**Comment by tkremenek** [comment]
@swift-ci test

**Comment by bitjammer** [comment]
Looks like John had some review additional comments. I'll address those now.

File: test/1_stdlib/WeakMirror.swift [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+struct StructHasObjCClassBoundExistential {
+  weak var weakProperty: NSObjectProtocol?
+}
+
+mirrors.test("class/NativeSwiftClassHasObjCWeakReference") {
+  let c1 = NativeSwiftClassHasWeak()
+  let nso = NSObject()
+  c1.weakProperty = nso
+  let classChild = _reflect(c1)[0].1.value
+  print(classChild)
```
Comment (gribozavr) [inline test/1_stdlib/WeakMirror.swift]: Could you use the new mirrors?  (`_reflect` should go away.)

Could you also add a stronger postcondition than just "print does not crash"?

**Comment by gparker42** [comment]
Test 1_stdlib/WeakMirror.swift is now failing on an asan bot:
https://smooshbase.apple.com/swiftci/view/Swift%20Build%20Czar/job/swift-whitney-incremental-asan-RDA/16/

I moved rdar://27348445 <rdar://27348445> back to Analyze.

**Comment by bitjammer** [comment]
Hm, I think I see why it's tripping it but it doesn't look immediately problematic to me. I think I can rearrange the code to make ASan happy, though.

**Pull request merged**